java-checks/src/main/java/org/sonar/java/checks/unused/UnusedLocalVariableCheck.java

```diff
@@ -19,128 +19,99 @@
-      addAssignment(((AssignmentExpressionTree) expression).variable());
-    } else if (expression.is(INCREMENT_KINDS)) {
-      addAssignment(((UnaryExpressionTree) expression).expression());
+  @Override
+  public void leaveNode(Tree tree) {
+    switch (tree.kind()) {
+      case IDENTIFIER:
+        IdentifierTree ident = (IdentifierTree) tree;
+        if (isRValue(ident)) {
+          unusedVariables.remove(ident.symbol());
+        }
+        break;
+      case VARIABLE:
+        addVariable((VariableTree) tree);
+        break;
+      case PATTERN_INSTANCE_OF:
+        addVariable(((PatternInstanceOfTree) tree).variable());
+        break;
+      default: // COMPILATION_UNIT
+        for (Map.Entry<Symbol, VariableTree> entry : unusedVariables.entrySet()) {
+          VariableTree variable = entry.getValue();
+          reportIssue(variable.simpleName(), "Remove this unused \"" + variable.symbol().name() + "\" local variable.");
+        }
+        break;
     }
   }
 
-  private void checkVariableUsages() {
-    for (VariableTree variableTree : variables) {
-      Symbol symbol = variableTree.symbol();
-      if (symbol.usages().size() == assignments.getOrDefault(symbol, Collections.emptyList()).size()) {
-        reportIssue(variableTree.simpleName(), "Remove this unused \"" + symbol.name() + "\" local variable.");
-      }
+  /**
+   * An identifier is being used as an r-value if it is not used as the left operand of an assignment nor as the
+   * operand of a stand-alone increment
+   */
+  private static boolean isRValue(IdentifierTree tree) {
+    if (tree.parent() instanceof AssignmentExpressionTree) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) tree.parent();
+      return assignment.variable() != tree;
     }
+    return !(tree.parent().is(INCREMENT_KINDS) && tree.parent().parent().is(Tree.Kind.EXPRESSION_STATEMENT));
   }
 
-  public void addVariables(List<StatementTree> statementTrees) {
-    UNRESOLVED_IDENTIFIERS_VISITOR.check(statementTrees);
-    for (StatementTree statementTree : statementTrees) {
-      if (statementTree.is(Tree.Kind.VARIABLE)) {
-        VariableTree variableTree = (VariableTree) statementTree;
-        if (!UNRESOLVED_IDENTIFIERS_VISITOR.isUnresolved(variableTree.simpleName().name())) {
-          variables.add(variableTree);
-        }
-      }
+  private void addVariable(VariableTree variable) {
+    if (isProperLocalVariable(variable) && !UNRESOLVED_IDENTIFIERS_VISITOR.isUnresolved(variable.simpleName().name())) {
+      unusedVariables.put(variable.symbol(), variable);
     }
   }
 
-  private void addAssignment(ExpressionTree variable) {
-    if (variable.is(Tree.Kind.IDENTIFIER)) {
-      addAssignment((IdentifierTree) variable);
-    }
+  private boolean isProperLocalVariable(VariableTree variable) {
+    return isLocalVariable(variable.symbol())
+      && !variable.parent().is(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR, Tree.Kind.LAMBDA_EXPRESSION, Tree.Kind.CATCH)
```

Comment:
By checking whether the parent is none of this types, we ensure that we target variables that are unused and can be effectively removed.
My question is (and I am not sure it is possible): Is there a way to express the types that **the parent could be** in a  concise way?
I think that would help with understanding  what kind of changes the rule can suggest

Reply:
I will rewrite this part to `&& !isParameter(variable.symbol()) && !isDefinedInTryCatch(variable.symbol())` using `JUtils.isParameter` and defining `isDefinedInTryCatch` as a private helper. That should make it easier to follow and also also be more robust in detecting parameters.